mne/decoding/time_gen.py

```diff
@@ -407,87 +465,114 @@ def _predict_time_loop(X, estimators, cv, slices, predict_mode,
-                # number of class.
-                if k == 0:
-                    # /!\ The CV may not be exhaustive. Thus, we need to fill
-                    # the prediction to an array whose length is similar to X
-                    # before it was rendered contiguous.
-                    this_ypred = np.empty((n_orig_epochs,) + y_pred_.shape[1:])
-                    y_pred.append(this_ypred)
-                y_pred[-1][test, ...] = y_pred_
-
-    if is_single_time_sample:
-        y_pred = [yp for yp in y_pred[0].transpose([1, 0, 2])]
+                # Initialize y_pred now we know its dimensionality
+                if y_pred is None:
+                    n_dim = y_pred_.shape[-1]
+                    y_pred = _init_ypred(n_train, n_test, n_orig_epochs, n_dim)
+                if vectorize_times:
+                    # Output predictions in different test_time columns
+                    y_pred[train_idx][test_time] = y_pred_
+                else:
+                    # Output predictions in a single test_time column
+                    y_pred[train_idx][test_idx] = y_pred_
+            elif predict_mode == 'cross-validation':
+                # Predict using the estimator corresponding to each fold
+                for (_, test), test_epoch, estimator in zip(
+                        splits, test_epochs, estimator_cv):
+                    if test.size == 0:
+                        continue
+                    y_pred_ = _predict(X_pred[test_epoch], [estimator],
+                                       vectorize_times=vectorize_times,
+                                       predict_method=predict_method)
+                    # Initialize y_pred now we know its dimensionality
+                    if y_pred is None:
+                        n_dim = y_pred_.shape[-1]
+                        y_pred = _init_ypred(n_train, n_test, n_orig_epochs,
+                                             n_dim)
+                    if vectorize_times:
+                        # Output predictions in different test_time columns
+                        y_pred[train_idx][test_time, test, ...] = y_pred_
+                    else:
+                        # Output predictions in a single test_time column
+                        y_pred[train_idx][test_idx, test, ...] = y_pred_
+
+    return y_pred
+
+
+def _init_ypred(n_train, n_test, n_orig_epochs, n_dim):
+    """
+    y_pred can only be initialized after the first prediction, because we
+    can't know whether it is a a categorical output or a set of
+    probabilistic estimates.
+    If all train time points have the same number of testing time
+    points, then y_pred is a matrix. Else it is an array of arrays.
+
+    Parameters
+    ----------
+    n_train : int
+        Number of training time point (i.e. estimators)
+    n_test : list of int
+        List of number of testing time points for each estimator.
+    n_orig_epochs : int
```

Comment:
Original number of predicted epochs before slice definition. Note
        that the number of epochs may have been cropped if the cross validation
        is not deterministic (e.g. with ShuffleSplit, we may only predict a
        subset of epochs).